lib/propolis/src/enlightenment/hyperv/overlay.rs

```diff
@@ -0,0 +1,754 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Support for hypervisor overlay pages.
+//!
+//! An _overlay page_ is a page of guest memory that "covers" a physical page in
+//! the guest's normal physical address space. The physical analogue might be a
+//! computer with a special one-page memory in addition to its regular RAM whose
+//! memory controller can be asked to redirect accesses to one page's worth of
+//! physical addresses to the auxiliary memory. The physical page at this
+//! address still exists in main memory, but it can't be addressed while the
+//! memory controller is applying the overlay.
+//!
+//! The virtualized case is similar, but there is only one physical memory (the
+//! host's memory), and the host plays the role of the memory controller,
+//! changing the guest's "view" of a particular GPA as overlays are applied and
+//! removed. One way to do this is to change the VM's nested page table entries
+//! (i.e., the translations from GPAs to HPAs) to redirect GPA accesses to a
+//! different host physical page. bhyve's memory management infrastructure
+//! doesn't currently support this, so this module emulates overlays in software
+//! by saving and restoring the contents of the various "layers" that exist at a
+//! particular GPA and making sure the active layer is present in the host
+//! physical page to which bhyve has mapped the relevant GPA.
+//!
+//! In general, the Hyper-V stack establishes overlays in response to a guest's
+//! request to place a particular overlay at a particular GPA. For example, the
+//! guest writes to the `HV_X64_MSR_HYPERCALL` MSR to provide the PFN of a page
+//! that the guest would like to see overlaid with a hypercall instruction page.
+//! This implies that overlays don't change while the guest's CPUs are paused;
+//! this has useful implications for the lifetimes of the structures in this
+//! module, which are described further below.
+//!
+//! TLFS section 5.2.1 describes the semantics of overlay pages in more detail.
+//! Importantly, the TLFS specifies that when multiple overlays exist for a
+//! single GPA, their logical ordering is implementation-defined (and not, for
+//! example, based on the order in which the overlays were requested).
+//!
+//! The TLFS doesn't explicitly specify the sizes of its overlay pages, but all
+//! the overlays this module cares about are 4 KiB, because the guest specifies
+//! their locations using 52-bit PFNs.
```

Comment:
I think it could be useful if this comment also included a high-level summary of our implementation of this, as well as discussing the HyperV concepts. In particular, there are a number of comments about the relationship between the lifecycles of the `OverlayManager` and the `OverlayPage`s it creates, but it would be nice if there was an introductory "theory of operation" type comment. I think it would be a bit easier to wrap ones head around those inline comments if there was an overview.

Reply:
...oh, silly me --- i later discovered that `OverlayManager` has a comment explaining this. personally, i might consider moving the struct definition up in the file so that it's encountered _before_ the code on `ManagerInner`  and `OverlayPage` that references this stuff...

Reply:
Done in 094bd27.